bundles/org.eclipse.core.runtime/src/org/eclipse/core/runtime/Plugin.java

```diff
@@ -526,6 +525,63 @@ public void setDebugging(boolean value) {
 		}
 	}
 
+	/**
+	 * Try to acquire a single service from the OSGi service factory (e.g.
+	 * DebugOptions, IWorkspace, IPreferenceService, ...)
+	 *
+	 * @param serviceType type of the service to acquire
+	 * @return an optional representing the acquired service or an empty optional if
+	 *         there is (currently) no service available or the plugin is shut down
+	 * @since 3.25
+	 */
+	public <T> Optional<T> lookup(Class<T> serviceType) {
```

Comment:
For this to work you need to access to the callers `BundleContext` then you need to track and get the service using that context, not the core.runtime context.  Then you need to track that bundle and react when the bundle is stopped so that you flush the usage of the service properly.

If you don't do this then you mess with the logic of the ServiceFactory that may depend on known what bundle is using its service.

Please have a look at `org.eclipse.core.runtime.ServiceCaller<Service>` and consider enhancing it with some of these ideas.  For example, adding a CompletableFuture.

Reply:
@tjwatson this does use the context of the plugin bundle not the` org.eclipse.core.runtime` BundleContext ...

Reply:
Ah, right!  Sorry I was thinking this was the `Platform` class.  Regardless I think the `ServiceCaller` should be considered instead of adding more API for accessing services.

Reply:
No problem, do you like to remove your change requesting review then?
`ServiceCaller` replicates unnecessary many code of of `ServiceTracker` so I never got why I should use this over the other.

Reply:
It depends on what that means to remove my change request review.  I am not convinced adding this new API to `Plugin` is necessary given that I think `ServiceCaller`can be used in most cases and we can consider enhancing it if we think some kind of future capability is needed.

I'm not sure what you mean by replicates code of `ServiceTracker`.  The advantage of `ServiceCaller` vs. `ServiceTracker` is simplicity of use over `ServiceTracker` and that it allows for `single` use service calls, as well as support for long-term usage of a service with many calls to the service.  The javadoc for `ServiceCaller` claims the following about `ServiceTracker`:

```
does not easily allow for lazy instantiation of the service instance.
```
Although I don't recall why that statement is true.  Perhaps @alblue can recall why he thought `ServiceTracker` made it harder to do lazy instantiation.

Reply:
> does not easily allow for lazy instantiation of the service instance.

Yep, I have had a discussion  with @alblue a while back, this claim is simply wrong and if one really wan't to have "call once" or similar all this could be done by delegating to a lazy fetching service tracker (I used one for example when implementing the AdapterFactory allowing OSGi services).

Reply:
> It depends on what that means to remove my change request review.

Currently you have requested a change on "For this to work you need to access to the callers BundleContext then you need to track and get the service using that context, not the core.runtime context.", you could resolve this conversation and simply do another review requesting an updated code location or comment.